### ISSUE TITLE
Guava dependency is problematic

### ISSUE THREAD

**Issue by marschall** [checkstyle/sonar-checkstyle#68]
I have a local experiment for a partial fix for #40. It reduces the `sonar-java-plugin` dependency to `java-squid`. Unfortunately the resulting artefact is not functional. During the build I get

```
[WARNING] com.google.guava:guava:jar:10.0.1:compile is provided by SonarQube plugin API and will not be packaged in your plugin
```

And at runtime I get

```
2017.02.09 09:37:47 ERROR web[o.s.s.p.Platform] Fail to stop server - ignored
java.lang.IllegalStateException: Cannot stop.  Current container state was: CONSTRUCTED
	at org.picocontainer.lifecycle.DefaultLifecycleState.stopping(DefaultLifecycleState.java:72) ~[picocontainer-2.15.jar:na]
	at org.picocontainer.DefaultPicoContainer.stop(DefaultPicoContainer.java:794) ~[picocontainer-2.15.jar:na]
	at org.picocontainer.DefaultPicoContainer.stop(DefaultPicoContainer.java:799) ~[picocontainer-2.15.jar:na]
	at org.picocontainer.DefaultPicoContainer.stop(DefaultPicoContainer.java:799) ~[picocontainer-2.15.jar:na]
	at org.sonar.core.platform.ComponentContainer.stopComponents(ComponentContainer.java:173) ~[sonar-core-5.6.5.jar:na]
	at org.sonar.core.platform.ComponentContainer.stopComponents(ComponentContainer.java:168) ~[sonar-core-5.6.5.jar:na]
	at org.sonar.server.platform.platformlevel.PlatformLevel.stop(PlatformLevel.java:93) ~[sonar-server-5.6.5.jar:na]
	at org.sonar.server.platform.Platform.stopLevel234Containers(Platform.java:240) ~[sonar-server-5.6.5.jar:na]
	at org.sonar.server.platform.Platform.doStop(Platform.java:269) ~[sonar-server-5.6.5.jar:na]
	at org.sonar.server.platform.PlatformServletContextListener.stopQuietly(PlatformServletContextListener.java:59) [sonar-server-5.6.5.jar:na]
	at org.sonar.server.platform.PlatformServletContextListener.contextInitialized(PlatformServletContextListener.java:52) [sonar-server-5.6.5.jar:na]
	at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4812) [tomcat-embed-core-8.0.32.jar:8.0.32]
	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5255) [tomcat-embed-core-8.0.32.jar:8.0.32]
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:147) [tomcat-embed-core-8.0.32.jar:8.0.32]
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1408) [tomcat-embed-core-8.0.32.jar:8.0.32]
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1398) [tomcat-embed-core-8.0.32.jar:8.0.32]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_121]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_121]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_121]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_121]
2017.02.09 09:37:47 ERROR web[o.a.c.c.C.[.[.[/]] Exception sending context initialized event to listener instance of class org.sonar.server.platform.PlatformServletContextListener
java.lang.IllegalStateException: Fail to load plugin Checkstyle [checkstyle]
	at org.sonar.server.plugins.ServerExtensionInstaller.installExtensions(ServerExtensionInstaller.java:74) ~[sonar-server-5.6.5.jar:na]
	at org.sonar.server.platform.platformlevel.PlatformLevel4.start(PlatformLevel4.java:653) ~[sonar-server-5.6.5.jar:na]
	at org.sonar.server.platform.Platform.start(Platform.java:216) ~[sonar-server-5.6.5.jar:na]
	at org.sonar.server.platform.Platform.startLevel34Containers(Platform.java:190) ~[sonar-server-5.6.5.jar:na]
	at org.sonar.server.platform.Platform.doStart(Platform.java:113) ~[sonar-server-5.6.5.jar:na]
	at org.sonar.server.platform.Platform.doStart(Platform.java:99) ~[sonar-server-5.6.5.jar:na]
	at org.sonar.server.platform.PlatformServletContextListener.contextInitialized(PlatformServletContextListener.java:44) ~[sonar-server-5.6.5.jar:na]
	at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4812) [tomcat-embed-core-8.0.32.jar:8.0.32]
	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5255) [tomcat-embed-core-8.0.32.jar:8.0.32]
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:147) [tomcat-embed-core-8.0.32.jar:8.0.32]
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1408) [tomcat-embed-core-8.0.32.jar:8.0.32]
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1398) [tomcat-embed-core-8.0.32.jar:8.0.32]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_121]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_121]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_121]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_121]
Caused by: java.lang.NoClassDefFoundError: com/google/common/collect/ImmutableList
	at org.sonar.plugins.checkstyle.CheckstylePlugin.getExtensions(CheckstylePlugin.java:38) ~[na:na]
	at org.sonar.api.SonarPlugin.define(SonarPlugin.java:51) ~[sonar-plugin-api-5.6.5.jar:na]
	at org.sonar.server.plugins.ServerExtensionInstaller.installExtensions(ServerExtensionInstaller.java:64) ~[sonar-server-5.6.5.jar:na]
	... 15 common frames omitted
Caused by: java.lang.ClassNotFoundException: com.google.common.collect.ImmutableList
	at org.sonar.classloader.ParentFirstStrategy.loadClass(ParentFirstStrategy.java:39) ~[sonar-classloader-1.0.jar:na]
	at org.sonar.classloader.ClassRealm.loadClass(ClassRealm.java:87) ~[sonar-classloader-1.0.jar:na]
	at org.sonar.classloader.ClassRealm.loadClass(ClassRealm.java:76) ~[sonar-classloader-1.0.jar:na]
	... 18 common frames omitted
```

I could not get the sonar-packaging-maven-plugin to include Guava.

The issue seems to be:
* before SonarQube 5.2 class loading of plugins was not isolated, plugins would simply see all classes provided by SonarQube. Since SonarQube relies on Guava this also includes Guava. The sonar-packaging-maven-plugin checks for this and does not include JARs which are provided by SonarQube.
* in SonarQube 5.2 the class loading was [changed](https://docs.sonarqube.org/display/DEV/Build+Plugin) all plugins are isolated now and see almost no classes from SonarQube (AFAIK only Sonar APIs and SLF4J but libraries used for implementation). Unfortunately the sonar-packaging-maven-plugin doesn't seem to have been updated to take this into account. It still assumes all kind of JARs will be provided by SonarQube when in fact this is no longer true.

I filed [sonar-packaging-maven-plugin#9](https://github.com/SonarSource/sonar-packaging-maven-plugin/issues/9).

**Comment by romani** [comment]
@marschall , do you want me to assign issue #40 to you.
If you want some issue to be assigned to you just make a post "I am on it". It will give signal to others to not touch issue

**Comment by marschall** [comment]
@romani I do not currently have plans beyond what I have done in #70 in terms of removing the dependency on Guava. What I plan to work on is my local prototype that reduces the `sonar-java-plugin` dependency to `java-squid` (part of #40).

**Comment by romani** [comment]
guava is not a problem anymore in recent version , fixed at 4.17, by https://github.com/checkstyle/sonar-checkstyle/issues/157

**Comment by rnveach** [comment]
Guava will be almost completely removed from main repo with PR https://github.com/checkstyle/checkstyle/pull/6504 .
All that will remain is classpath scanning from guava.
